ansible_base/lib/utils/views.py

```diff
@@ -0,0 +1,105 @@
+import importlib
+import logging
+import time
+
+from django.conf import settings
+from django.db import connection
+from django.utils.translation import gettext_lazy as _
+from rest_framework.exceptions import AuthenticationFailed, ParseError, PermissionDenied, UnsupportedMediaType
+from rest_framework.views import APIView
+
+from ansible_base.lib.utils.settings import get_function_from_setting, get_setting
+
+logger = logging.getLogger('ansible_base.lib.utils.views')
+
+
+class AnsibleBaseView(APIView):
+    def initialize_request(self, request, *args, **kwargs):
+        """
+        Store the Django REST Framework Request object as an attribute on the
+        normal Django request, store time the request started.
+        """
+        self.time_started = time.time()
+        if get_setting('SQL_DEBUG', False):
+            self.queries_before = len(connection.queries)
+
+        # If there are any custom headers in REMOTE_HOST_HEADERS, make sure
+        # they respect the allowed proxy list
+        proxy_ip_allowed_list = get_setting('PROXY_IP_ALLOWED_LIST', None)
+        if proxy_ip_allowed_list and all(
+            [
+                request.environ.get('REMOTE_ADDR') not in proxy_ip_allowed_list,
+                request.environ.get('REMOTE_HOST') not in proxy_ip_allowed_list,
+            ]
+        ):
+            for custom_header in get_setting('REMOTE_HOST_HEADERS', None):
+                if custom_header.startswith('HTTP_'):
+                    request.environ.pop(custom_header, None)
+
+        drf_request = super().initialize_request(request, *args, **kwargs)
+        request.drf_request = drf_request
+        try:
+            request.drf_request_user = getattr(drf_request, 'user', False)
+        except AuthenticationFailed:
+            request.drf_request_user = None
+        except (PermissionDenied, ParseError) as exc:
+            request.drf_request_user = None
+            self.__init_request_error__ = exc
+        except UnsupportedMediaType as exc:
+            exc.detail = _(
+                'You did not use correct Content-Type in your HTTP request. If you are using our REST API, the Content-Type must be application/json'
+            )
+            self.__init_request_error__ = exc
+        return drf_request
+
+    def finalize_response(self, request, response, *args, **kwargs):
+        response = super().finalize_response(request, response, *args, **kwargs)
+
+        if request.user.is_authenticated:
```

Comment:
I saw the test failures about the request object not having a user, and thought it might be some fatal flaw with the crum library. But no, seems it's just because the mock_request fixture doesn't have a user.

Reply:
Yea, the tests need a full rewrite based on the changes I've made. Working on that currently.